lxd/events/connections.go

```diff
@@ -128,3 +132,103 @@ func (e *websockListenerConnection) WriteJSON(event any) error {
 
 	return e.Conn.WriteJSON(event)
 }
+
+// NewStreamListenerConnection returns a new http stream listener connection.
+func NewStreamListenerConnection(connection net.Conn) EventListenerConnection {
+	// Send HTTP response to let the client know what to expect.
+	// This is only sent once, and is followed by events.
+	//
+	// The X-Content-Type-Options response HTTP header is a marker used by the server to indicate
+	// that the MIME types advertised in the Content-Type headers should be followed and not be
+	// changed. The header allows you to avoid MIME type sniffing by saying that the MIME types are
+	// deliberately configured.
+	io.WriteString(connection, `HTTP/1.1 200 OK
+Connection: keep-alive
+Content-Type: application/json
+X-Content-Type-Options: nosniff
+
+`)
+
+	return &streamListenerConnection{
+		Conn: connection,
+	}
+}
+
+func (e *streamListenerConnection) Reader(ctx context.Context, recvFunc EventHandler) {
+	ctx, cancelFunc := context.WithCancel(ctx)
+
+	close := func() {
+		e.lock.Lock()
+		defer e.lock.Unlock()
+
+		if ctx.Err() != nil {
+			return
+		}
+
+		e.Close()
+		cancelFunc()
+	}
+
+	defer close()
+
+	pingInterval := time.Second * 10
+
+	// Start reader from client.
+	go func() {
+		defer close()
+
+		buf := make([]byte, 1)
+
+		for {
+			err := e.SetReadDeadline(time.Now().Add(5 * time.Second))
+			if err != nil {
+				logger.Warn("Failed setting read deadline", logger.Ctx{"local": e.LocalAddr(), "remote": e.RemoteAddr(), "err": err})
+				return
+			}
+
+			// This is used to determine whether the client has terminated.
+			_, err = e.Read(buf)
+			if err != nil && errors.Is(err, io.EOF) {
+				return
+			}
+		}
+	}()
+
+	for {
+		if ctx.Err() != nil {
+			return
+		}
+
+		select {
+		case <-time.After(pingInterval):
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (e *streamListenerConnection) WriteJSON(event any) error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	err := e.SetWriteDeadline(time.Now().Add(5 * (time.Second)))
+	if err != nil {
+		return fmt.Errorf("Failed setting write deadline: %w", err)
+	}
+
+	data, err := json.Marshal(event)
```

Comment:
You should be able to use a json encoder to avoid the intermediate representation of the data here.

https://pkg.go.dev/encoding/json#NewEncoder

As the encoded json object is delimited with a newline:

https://cs.opensource.google/go/go/+/refs/tags/go1.18.2:src/encoding/json/stream.go;l=217